**EJERCICIO EN CLASE 18/10/2024**

In [2]:
import pandas as pd

data = pd.read_csv('datos_ejercicio_ventas.csv', delimiter=';')
data.head()

COUNTRY        SUBBRAND  YEAR  MONTH     SCENARIO FORECAST  \
0       Portugal     Lipton (L3)  2023     12  AI_forecast  AI_P02F   
1  Great Britain     Lipton (L3)  2023     12  AI_forecast  AI_P10F   
2          Spain  Pepsi Max (L3)  2023     12  AI_forecast  AI_P09F   
3  Great Britain        7up (L3)  2024     12  AI_forecast  AI_P10F   
4        Hungary     Lipton (L3)  2023      9  AI_forecast  AI_P03F   

   FORECAST_YEAR       AMOUNT  
0         2023.0  7,54356E+15  
1         2023.0  5,60031E+15  
2         2023.0   8,8502E+15  
3         2023.0  3,63225E+15  
4         2023.0  3,96176E+16

In [3]:
data.shape

(18666, 8)

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18666 entries, 0 to 18665
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   COUNTRY        18666 non-null  object 
 1   SUBBRAND       18666 non-null  object 
 2   YEAR           18666 non-null  int64  
 3   MONTH          18666 non-null  int64  
 4   SCENARIO       18666 non-null  object 
 5   FORECAST       17766 non-null  object 
 6   FORECAST_YEAR  17766 non-null  float64
 7   AMOUNT         18666 non-null  object 
dtypes: float64(1), int64(2), object(5)
memory usage: 1.1+ MB


DATASET:
CADA REGISTRO REPRESENTA LA CANTIDAD VENDIDA DE UN PRODUCTO EN UN PAÍS Y EN UN INSTANTE TEMPORAL (DADO EN MESES). ESTA CANTIDAD PUEDE SER UN DATO REAL O UNA PREDICCIÓN HECHA.
* 'COUNTRY': STRING, PAÍS, VARIABLE CATEGÓRICA.
* 'SUBBRAND': STRING, PRODUCTO, VARIABLE CATEGÓRICA.
* 'YEAR': INT, AÑO DEL VALOR, VARIABLE DISCRETA.
* 'MONTH': INT, MES DEL VALOR, VARIABLE DISCRETA.
* 'SCENARIO': STRING, INDICA SI EL VALOR ES REAL O UNA PREDICCIÓN, VARIABLE CATEGÓRICA.
* 'FORECAST': STRING, SI ES UNA PREDICCIÓN, INDICA EL PERÍODO DESDE EL QUE SE HACE LA PREDICCIÓN. MÁS CONCRETAMENTE, 'P0X' indica que se está en el final del mes X-1. LO INDICA A MES VENCIDO. 'P02' ES FINALES DE ENERO / PRINCIPIOS DE FEBRERO. VARIABLE CATEGORICA
* 'FORECAST_YEAR': FLOAT, EL AÑO DESDE EL QUE SE ESTÁ HACIENDO LA PREVISIÓN, VARIABLE DISCRETA.
* 'AMOUNT': STRING, AUNQUE REPRESENTA UN NÚMERO. E+ALGO SE REFIERE A E A LA ALGO. REPRESENTA UNA VARIABLE CONTINUA.

**NULLS EN 'FORECAST' Y 'FORECAST_YEAR', ADEMÁS EL MISMO Nº DE NULLS EN CADA UNA** 

In [5]:
data.isna().sum()

COUNTRY            0
SUBBRAND           0
YEAR               0
MONTH              0
SCENARIO           0
FORECAST         900
FORECAST_YEAR    900
AMOUNT             0
dtype: int64

In [6]:
data[(data['FORECAST'].isna()) & (data['FORECAST_YEAR'].isna())]

COUNTRY            SUBBRAND  YEAR  MONTH SCENARIO FORECAST  \
277         Portugal      Pepsi Max (L3)  2023     10   actual      NaN   
278         Portugal            7up (L3)  2023      3   actual      NaN   
279         Portugal            7up (L3)  2023     10   actual      NaN   
280    Great Britain       7up Free (L3)  2023     10   actual      NaN   
281           Norway  Pepsi Regular (L3)  2023     10   actual      NaN   
...              ...                 ...   ...    ...      ...      ...   
18595  Great Britain         Lipton (L3)  2023      9   actual      NaN   
18596          Italy  Pepsi Regular (L3)  2023      9   actual      NaN   
18597  Great Britain       7up Free (L3)  2023      9   actual      NaN   
18598  Great Britain   Mountain Dew (L3)  2023      9   actual      NaN   
18599          Spain       7up Free (L3)  2023      9   actual      NaN   

       FORECAST_YEAR       AMOUNT  
277              NaN  1,88595E+16  
278              NaN  2,93497E+15  
279              NaN  3,48447E+16  
280              NaN  1,17255E+16  
281              NaN  3,78486E+15  
...              ...          ...  
18595            NaN  8,49263E+15  
18596            NaN  1,56257E+16  
18597            NaN  1,29323E+15  
18598            NaN  2,06646E+16  
18599            NaN  7,58888E+14  

[900 rows x 8 columns]

**TODOS LOS NAN'S DE 'FORECAST' SON TAMBIÉN TODOS LOS NAN'S DE 'FORECAST_YEAR'**

In [8]:
data[data['SCENARIO'] == 'actual']

COUNTRY            SUBBRAND  YEAR  MONTH SCENARIO FORECAST  \
277         Portugal      Pepsi Max (L3)  2023     10   actual      NaN   
278         Portugal            7up (L3)  2023      3   actual      NaN   
279         Portugal            7up (L3)  2023     10   actual      NaN   
280    Great Britain       7up Free (L3)  2023     10   actual      NaN   
281           Norway  Pepsi Regular (L3)  2023     10   actual      NaN   
...              ...                 ...   ...    ...      ...      ...   
18595  Great Britain         Lipton (L3)  2023      9   actual      NaN   
18596          Italy  Pepsi Regular (L3)  2023      9   actual      NaN   
18597  Great Britain       7up Free (L3)  2023      9   actual      NaN   
18598  Great Britain   Mountain Dew (L3)  2023      9   actual      NaN   
18599          Spain       7up Free (L3)  2023      9   actual      NaN   

       FORECAST_YEAR       AMOUNT  
277              NaN  1,88595E+16  
278              NaN  2,93497E+15  
279              NaN  3,48447E+16  
280              NaN  1,17255E+16  
281              NaN  3,78486E+15  
...              ...          ...  
18595            NaN  8,49263E+15  
18596            NaN  1,56257E+16  
18597            NaN  1,29323E+15  
18598            NaN  2,06646E+16  
18599            NaN  7,58888E+14  

[900 rows x 8 columns]

**TODOS LOS NAN'S SE DAN CUANDO EL DATO ES REAL. ESTO TIENE SENTIDO, YA QUE SI TENGO EL DATO REAL, NO HAY PREDICCIÓN**

**COSAS QUE QUIERE VER EL PROFE**:
1. Nº ACTUALS Y FORECAST.
2. HORIZONTE DE PREVISIÓN. PISTA: ES EL MISMO SIEMPRE.
3. Nº PAÍSES, PRODUCTOS.
4. HISTÓRICO DE DATOS DE ACTUALS / FORECAST.
5. Nº DE FORECAST DISTINTOS.

1. Nº ACTUALS Y FORECAST.

In [12]:
data[data['SCENARIO'] == 'actual'].shape[0]

900

HAY 900 DATOS REALES.

In [14]:
data[data['SCENARIO'] == 'AI_forecast'].shape[0]

17766

HAY 17766 VALORES PREDICHOS.

2. HORIZONTE DE PREVISIÓN.

El instante en el que se predice para una marca de un país viene dado por la combinación de 'FORECAST' y 'FORECAST_YEAR'. Para ver cuál es el horizonte de previsión para cada predicción hecha, basta con ver cuántas filas hay para cada combinación de ('FORECAST', 'FORECAST_YEAR', 'COUNTRY', 'SUBBRAND').

In [17]:
data.groupby(['FORECAST', 'FORECAST_YEAR', 'COUNTRY', 'SUBBRAND']).size().head(50)

FORECAST  FORECAST_YEAR  COUNTRY  SUBBRAND          
AI_P02F   2023.0         Czech    7up (L3)              18
                                  Lipton (L3)           18
                                  Mountain Dew (L3)     18
                                  Pepsi Max (L3)        18
                                  Pepsi Regular (L3)    36
                                                        ..
AI_PF     2023.0         Spain    7up (L3)              18
                                  7up Free (L3)         18
                                  Lipton (L3)           18
                                  Pepsi Max (L3)        18
                                  Pepsi Regular (L3)    18
Length: 564, dtype: int64

In [16]:
import pandas as pd
import numpy as np

# Agrupar por las columnas necesarias y contar el número de ocurrencias
horizontes_distintos = data.groupby(['FORECAST', 'FORECAST_YEAR', 'COUNTRY', 'SUBBRAND']).size()

# Obtener los valores únicos de los tamaños de los grupos
horizontes_unicos = horizontes_distintos.unique()

# Mostrar los horizontes únicos
print(horizontes_unicos)

[ 18  36  54  72  90 180]
